File: rust/src/record.rs
use std::collections::HashSet;

use crate::block::{
    unmarshal_chunk_header, unmarshal_descriptor_header, ValueChunkHeader,
};

use crate::device::BlockDevice;
use crate::error::Result;
use crate::format::*;

/// A fully validated record ready for returning to callers.
pub struct CompleteRecord {
    pub key: Vec<u8>,
    pub generation: u32,
    pub descriptor_block: u32,
    pub header: crate::block::RecordDescriptorHeader,
    /// Block indices of value chunk blocks (chunk_index >= 1).
    pub chunk_blocks: Vec<u32>,
    pub value: Vec<u8>,
}

/// Read and validate the record whose descriptor is at `descriptor_block`.
/// Returns `Ok(None)` when the record is incomplete or corrupt.
pub fn verify_and_read_record<D: BlockDevice>(
    dev: &mut D,
    descriptor_block: u32,
) -> Result<Option<CompleteRecord>> {
    let block_size = dev.block_size();
    let mut buf = vec![0u8; block_size as usize];

    dev.read_block(descriptor_block, &mut buf)?;

    if buf[0] != BLOCK_TYPE_RECORD_DESCRIPTOR {
        return Ok(None);
    }

    let mut hdr = crate::block::RecordDescriptorHeader::default();
    if !unmarshal_descriptor_header(&buf, &mut hdr) {
        return Ok(None);
    }
    if hdr.header_size as u32 != RECORD_DESCRIPTOR_HEADER_SIZE {
        return Ok(None);
    }
    if hdr.key_size as u32 >= block_size - RECORD_DESCRIPTOR_HEADER_SIZE {
        return Ok(None);
    }
    let first_payload_cap = block_size - RECORD_DESCRIPTOR_HEADER_SIZE - hdr.key_size as u32;
    if hdr.first_payload_size > first_payload_cap {
        return Ok(None);
    }
    if hdr.chunk_count == 0 {
        return Ok(None);
    }

    // Read key bytes
    let key_start = RECORD_DESCRIPTOR_HEADER_SIZE as usize;
    let key_end = key_start + hdr.key_size as usize;
    let key = buf[key_start..key_end].to_vec();

    // Read first value payload — OOM guard
    let max_dev_bytes = block_size as u64 * dev.block_count() as u64;
    let cap_value = (hdr.total_size as u64).min(max_dev_bytes) as usize;
    let cap_chunks = (hdr.chunk_count.saturating_sub(1)).min(dev.block_count()) as usize;

    let payload_start = key_end;
    let payload_end = payload_start + hdr.first_payload_size as usize;
    let mut value = Vec::with_capacity(cap_value);
    value.extend_from_slice(&buf[payload_start..payload_end]);

    let mut rec = CompleteRecord {
        key,
        generation: hdr.generation,
        descriptor_block,
        header: hdr.clone(),
        chunk_blocks: Vec::with_capacity(cap_chunks),
        value,
    };

    if hdr.chunk_count == 1 {
        if hdr.total_size != hdr.first_payload_size || hdr.next_chunk != NULL_BLOCK_INDEX {
            return Ok(None);
        }
        return Ok(Some(rec));
    }

    if hdr.next_chunk == NULL_BLOCK_INDEX {
        return Ok(None);
    }

    let mut chunk_buf = vec![0u8; block_size as usize];
    let mut next = hdr.next_chunk;
    let mut expected_idx: u32 = 1;
    let mut visited: HashSet<u32> = HashSet::new();
    visited.insert(descriptor_block);

    loop {
        if next == NULL_BLOCK_INDEX {
            break;
        }
        if next >= dev.block_count() || visited.contains(&next) {
            return Ok(None);
        }
        visited.insert(next);

        dev.read_block(next, &mut chunk_buf)?;

        if chunk_buf[0] != BLOCK_TYPE_VALUE_CHUNK {
            return Ok(None);
        }

        let mut ch = ValueChunkHeader::default();
        if !unmarshal_chunk_header(&chunk_buf, &mut ch) {
            return Ok(None);
        }
        if ch.header_size as u32 != VALUE_CHUNK_HEADER_SIZE {
            return Ok(None);
        }
        if ch.owner_descriptor != descriptor_block {
            return Ok(None);
        }
        if ch.chunk_index != expected_idx {
            return Ok(None);
        }
        if ch.payload_size > block_size - VALUE_CHUNK_HEADER_SIZE {
            return Ok(None);
        }

        let pstart = VALUE_CHUNK_HEADER_SIZE as usize;
        let pend = pstart + ch.payload_size as usize;
        rec.value.extend_from_slice(&chunk_buf[pstart..pend]);
        rec.chunk_blocks.push(next);

        expected_idx += 1;
        next = ch.next_chunk;
    }

    if expected_idx != hdr.chunk_count {
        return Ok(None);
    }
    if rec.value.len() as u32 != hdr.total_size {
        return Ok(None);
    }

    Ok(Some(rec))
}

/// Calculate the number of blocks needed to store key + value.
pub fn needed_blocks(key: &[u8], value: &[u8], block_size: u32) -> u32 {
    let first_cap = block_size - RECORD_DESCRIPTOR_HEADER_SIZE - key.len() as u32;
    let chunk_cap = block_size - VALUE_CHUNK_HEADER_SIZE;
    let vlen = value.len() as u32;
    if vlen <= first_cap {
        return 1;
    }
    let remaining = vlen - first_cap;
    let extra = (remaining + chunk_cap - 1) / chunk_cap;
    1 + extra
}

File: go/record.go
package embedkv

// neededBlocks returns the number of blocks required to store key+value.
func neededBlocks(key, value []byte, blockSize uint32) uint32 {
	firstCap := firstPayloadCapacity(blockSize, uint32(len(key)))
	chunkCap := chunkPayloadCapacity(blockSize)
	vlen := uint32(len(value))
	if vlen <= firstCap {
		return 1
	}
	remaining := vlen - firstCap
	extra := (remaining + chunkCap - 1) / chunkCap
	return 1 + extra
}

// completeRecord holds the fully validated content of a readable record.
type completeRecord struct {
	key             []byte
	generation      uint32
	descriptorBlock uint32
	header          RecordDescriptorHeader
	chunkBlocks     []uint32 // block indices of value chunk blocks (chunk_index >= 1)
	value           []byte
}

// verifyAndReadRecord reads and validates the record whose descriptor is at
// descriptorBlock. Returns nil (no error) when the record is incomplete or corrupt.
func verifyAndReadRecord(dev BlockDevice, descriptorBlock uint32) (*completeRecord, error) {
	blockSize := dev.BlockSize()
	buf := make([]byte, blockSize)

	if err := dev.ReadBlock(descriptorBlock, buf); err != nil {
		return nil, err
	}
	if buf[0] != BlockTypeRecordDescriptor {
		return nil, nil
	}

	var hdr RecordDescriptorHeader
	if !unmarshalDescriptorHeader(buf, &hdr) {
		return nil, nil
	}
	if uint32(hdr.HeaderSize) != RecordDescriptorHeaderSize {
		return nil, nil
	}
	if uint32(hdr.KeySize) > blockSize-RecordDescriptorHeaderSize-BlockCRCSize {
		return nil, nil
	}
	firstPayloadCap := firstPayloadCapacity(blockSize, uint32(hdr.KeySize))
	if hdr.FirstPayloadSize > firstPayloadCap {
		return nil, nil
	}
	if hdr.ChunkCount == 0 {
		return nil, nil
	}

	// Read key bytes
	keyStart := RecordDescriptorHeaderSize
	keyEnd := keyStart + uint32(hdr.KeySize)
	key := make([]byte, hdr.KeySize)
	copy(key, buf[keyStart:keyEnd])

	// Read first value payload
	payloadStart := keyEnd
	payloadEnd := payloadStart + hdr.FirstPayloadSize
	// Cap initial slice capacities to the physical device size to prevent OOM
	// when headers contain extreme values but a valid CRC (e.g., fuzz inputs).
	// Actual bytes appended via append are bounded by block content, not header fields.
	maxDevBytes := uint64(blockSize) * uint64(dev.BlockCount())
	capValue := uint64(hdr.TotalSize)
	if capValue > maxDevBytes {
		capValue = maxDevBytes
	}
	// ChunkCount >= 1 already verified above; cap at block count (max reachable chain length)
	capChunks := uint64(hdr.ChunkCount - 1)
	if capChunks > uint64(dev.BlockCount()) {
		capChunks = uint64(dev.BlockCount())
	}

	value := make([]byte, 0, capValue)
	value = append(value, buf[payloadStart:payloadEnd]...)

	rec := &completeRecord{
		key:             key,
		generation:      hdr.Generation,
		descriptorBlock: descriptorBlock,
		header:          hdr,
		chunkBlocks:     make([]uint32, 0, capChunks),
		value:           value,
	}

	if hdr.ChunkCount == 1 {
		if hdr.TotalSize != hdr.FirstPayloadSize || hdr.NextChunk != NullBlockIndex {
			return nil, nil
		}
		return rec, nil
	}

	if hdr.NextChunk == NullBlockIndex {
		return nil, nil
	}

	chunkBuf := make([]byte, blockSize)
	next := hdr.NextChunk
	expectedIdx := uint32(1)
	visited := make(map[uint32]bool)
	visited[descriptorBlock] = true

	for next != NullBlockIndex {
		if next >= dev.BlockCount() || visited[next] {
			return nil, nil
		}
		visited[next] = true

		if err := dev.ReadBlock(next, chunkBuf); err != nil {
			return nil, err
		}
		if chunkBuf[0] != BlockTypeValueChunk {
			return nil, nil
		}

		var ch ValueChunkHeader
		if !unmarshalChunkHeader(chunkBuf, &ch) {
			return nil, nil
		}
		if uint32(ch.HeaderSize) != ValueChunkHeaderSize {
			return nil, nil
		}
		if ch.OwnerDescriptor != descriptorBlock {
			return nil, nil
		}
		if ch.ChunkIndex != expectedIdx {
			return nil, nil
		}
		if ch.PayloadSize > chunkPayloadCapacity(blockSize) {
			return nil, nil
		}

		rec.value = append(rec.value, chunkBuf[ValueChunkHeaderSize:ValueChunkHeaderSize+ch.PayloadSize]...)
		rec.chunkBlocks = append(rec.chunkBlocks, next)

		expectedIdx++
		next = ch.NextChunk
	}

	if expectedIdx != hdr.ChunkCount {
		return nil, nil
	}
	if uint32(len(rec.value)) != hdr.TotalSize {
		return nil, nil
	}

	return rec, nil
}
